### PR TITLE
increase cachi2 coverage

### PR DESCRIPTION
**Pull request by slimreaper35**
# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

**Comment by slimreaper35** [comment]
https://coveralls.io/builds/66115333

File: pyproject.toml [hunk 1]
```diff
@@ -82,3 +82,15 @@ disallow_untyped_decorators = true
 [[tool.mypy.overrides]]
 module = ["semver"]
 ignore_missing_imports = true
+
+[tool.coverage.report]
+skip_covered = true
+show_missing = true
+fail_under = 80
```
[the remark this answers is not in the document]
Comment (slimreaper35) [reply pyproject.toml:89]: How significant is this number by the way? If we set `fail_udner` to an even higher number like 90 or 95.

File: cachi2/core/rooted_path.py [hunk 2]
```diff
@@ -70,9 +70,8 @@ def subpath_from_root(self) -> Path:
         return self._path.relative_to(self._root)
 
     def __eq__(self, other: object) -> bool:
-        # Note: __eq__ is implemented mainly for unit tests that do assert_called_with() or similar
         if not isinstance(other, RootedPath):
-            return NotImplemented
+            raise NotImplementedError(f"__eq__ not implemented for {type(other).__name__}")
```
[the remark this answers is not in the document]
Comment (slimreaper35) [reply cachi2/core/rooted_path.py:74]: Well, I thought, that raising an exception wouldn't be better than just returning **False** while comparing _RootedPath_ with _pathlib.Path_ for example. I might save some debugging time.

File: .github/workflows/gating.yaml [hunk 3]
```diff
@@ -39,7 +41,7 @@ jobs:
       run: |
         pip3 install --upgrade pip
         pip3 install --upgrade setuptools
-        pip3 install --upgrade coveralls==3.2.0
+        pip3 install --upgrade coveralls
```
[the remark this answers is not in the document]
Comment (slimreaper35) [reply .github/workflows/gating.yaml:44]: I will take another look at this.

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ben-alkov) [reply pyproject.toml]: Current coverage is >97% (https://coveralls.io/github/containerbuildsystem/cachi2) - there is *no* reason why this number should be this low.

File: cachi2/core/rooted_path.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/rooted_path.py:74]: >  I thought, that raising an exception wouldn't be better than just returning False while comparing RootedPath with pathlib.Path for example. I might save some debugging time.

So I read the docs on `NotImplementedError` exception and I have to retract on my previous answer. The intended use case is for base classes and abstract methods, which makes sense during development. However, while you're perfectly fine to use it anyway you wish, there are a couple of things to consider which I believe suggest against using it in this case:

1. Deliberately raising an exception on a production code instead of wrapping it by a nice user-friendly error is an antipattern which means `NotImplemented` is actually a better choice here
2. Pathlib itself uses `NotImplemented` instead of an exception: https://github.com/python/cpython/blob/d308d33e098d8e176f1e5169225d3cf800ed6aa1/Lib/pathlib/__init__.py#L203

File: pyproject.toml [hunk 4]
```diff
@@ -82,3 +82,15 @@ disallow_untyped_decorators = true
...
+exclude_lines = [
+  "def __repr__",
+  "if __name__ == .__main__.:",
+  "if TYPE_CHECKING:",
+  "raise AssertionError",
+  "raise NotImplementedError",
```
[the remark this answers is not in the document]
Comment (eskultety) [reply pyproject.toml:94]: Considering dropping `NotImplementedError` completely, this hunk would need to be adjusted accordingly.

File: cachi2/core/rooted_path.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (slimreaper35) [reply cachi2/core/rooted_path.py:74]: That's a good argument. I just kind of don't understand what the extra value of `NotImplemented`.
```
In [1]: from cachi2.core.rooted_path import RootedPath

In [2]: x = RootedPath("/home/cachi2")

In [3]: from pathlib import Path

In [4]: y = Path("/home/cachi2")

In [5]: x == y
Out[5]: False
```

File: .github/workflows/gating.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (slimreaper35) [reply .github/workflows/gating.yaml]: Well, I got an error with pinned version:

 `coverage.misc.CoverageException: Can't read 'pyproject.toml' without TOML support. Install with [toml] extra`

File: cachi2/core/rooted_path.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/rooted_path.py:74]: Well, you will need to go through [1] and [2] yourself if you want deep understanding ;), but the gist here is that `NotImplemented` is a special Python builtin type which is used solely with rich comparisons (i.e. usage of methods `__eq__`, `__lt__`, `__gt__`, etc. as a replacement of the old mechanism implemented by `__cmp__`) and nothing else, in other words, you wouldn't use `NotImplemented` for a random function/method that you declared, but deferred the implementation  (for that you'd use an exception). IIUC `NotImplemented` serves the purpose of chain hierarchy of fallback behaviours ultimately leading to a `TypeError` which I understand you'd be kind of after from your example.

What I think is important to understand here is that while ultimately you'd still get a `TypeError`, Python is trying hard to defer that logic to the worst case scenario and tries other possibilities first, ridding you of the need to put error recovering logic absolutely everywhere.

[1] https://docs.python.org/3/library/constants.html#NotImplemented
[2] https://docs.python.org/3/reference/datamodel.html#object.__eq__

File: .github/workflows/gating.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (eskultety) [reply .github/workflows/gating.yaml]: > Well, I got an error with the pinned version:
>
> `coverage.misc.CoverageException: Can't read 'pyproject.toml' without TOML support. Install with [toml] extra`

This has nothing to do with pinning/not pinning the version though. The reason why you're seeing the error is caused by your own change in this PR:
```
diff --git a/.github/workflows/gating.yaml b/.github/workflows/gating.yaml
index 3ade3e65..1b9bf7a6 100644
--- a/.github/workflows/gating.yaml
+++ b/.github/workflows/gating.yaml
@@ -28,7 +28,9 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install tox tox-gh-actions
+        pip install tox
+        pip install tox-gh-actions
+        pip install coverage
```

In the pre-coveralls step you install the latest coverage there is, which is >=7, however, then you install an unpinned version of `coveralls` which, since it has an upper boundary on coverage<7 it will downgrade it to coverage==6.X [which moved away from `setup.cfg` to `pyproject.toml`](https://github.com/nedbat/coveragepy/releases/tag/6.5.0) and so now it requires you to install the extra toml dependency [it has](https://github.com/nedbat/coveragepy/blob/3d57a072ec6073988b4c775a9cffd431ef63ab5f/setup.py#L100). Moreover, if you look at our `requirements-extras.txt` it's exactly what we're doing already for local dev/test envs: 
```
requirements-extras.txt=258=coverage[toml]==7.4.3 \
requirements-extras.txt:311:    # via pytest-cov
```

That said, I dug deeper into testing this and it turned out that what I described in https://github.com/containerbuildsystem/cachi2/pull/481#discussion_r1519822031 happens also with `coveralls==3.2.0` which then downgrades `coverage` to v5.5 with the same error. Looks like this all boils down to the fact that `coveralls` is an abandonware now: https://github.com/TheKevJames/coveralls-python/issues/393 and we should ultimately look for a different solution if we wish to keep reporting the coverage online (do we?). Since the behaviour of `coveralls` installation wrt/ `coverage` seems to be consistently broken no matter what version I am okay if you proceed with dropping the version pin, but please do it in a separate commit with and explanation - I think I've provided enough material to reference in the commit message :) .

File: .github/workflows/gating.yaml [hunk 5]
```diff
@@ -28,7 +28,9 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install tox tox-gh-actions
+        pip install tox
+        pip install tox-gh-actions
+        pip install coverage
```
Comment (eskultety) [inline .github/workflows/gating.yaml:33]: github action changes should not be part of this commit. This commit is solely about moving configuration from .coveragerc to pyproject.toml

File: .coveragerc [hunk 6]
```diff
@@ -3,6 +3,7 @@ skip_covered = True
 show_missing = True
 fail_under = 75
 exclude_lines =
-    pragma: no cover
-    if __name__ == .__main__.:
     def __repr__
+    if __name__ == .__main__.:
+    if TYPE_CHECKING:
+    raise NotImplementedError
```
Comment (eskultety) [inline .coveragerc]: You're still have `NotImplementedError` here even though we're not using it.

File: pyproject.toml [hunk 7]
```diff
@@ -86,7 +86,7 @@ ignore_missing_imports = true
 [tool.coverage.report]
```
Comment (eskultety) [inline pyproject.toml:86]: **commit message:**

It would be nice to have some, even brief, rationale behind this in the commit message rather than leaving it empty.

Also: 
`s/coverage fail_under value to 90/code coverage to 90%`

File: .coveragerc [hunk 8]
```diff
@@ -3,6 +3,7 @@ skip_covered = True
 show_missing = True
 fail_under = 75
 exclude_lines =
-    pragma: no cover
-    if __name__ == .__main__.:
     def __repr__
```
Comment (eskultety) [inline .coveragerc]: **commit message:**
With the latest updates to this PR this specific commit subject is rather confusing as you're by no means increasing coverage, merely adding a few more patterns to ignore during the coverage calculation.

File: .coveragerc [hunk 9]
```diff
@@ -3,6 +3,7 @@ skip_covered = True
...
+    if __name__ == .__main__.:
```
Comment (eskultety) [inline .coveragerc]: Why did this line move? Did you sort this alphabetically manually or accidentally with some tool? Anyway, this line change is totally unnecessary especially since the file is going away in favour of pyproject.toml later in this series.

File: .github/workflows/gating.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ben-alkov) [reply .github/workflows/gating.yaml]: > if we wish to keep reporting the coverage online

IMO, it's more about having *some* report, vs. having to dig around in GH action logs (I'm still trying to get over the fact that we don't have useful reporting from pytest... (pytest-html, anyone?))

[the remark this answers is not in the document]
Comment (slimreaper35) [reply .github/workflows/gating.yaml:43]: Should I fill out a story on JIRA ?

[the remark this answers is not in the document]
Comment (eskultety) [reply .github/workflows/gating.yaml:43]: > Should I fill out a story on JIRA ?

I think we can keep this one contained solely in the upstream issue tracker rather than internally, but yes, we should track the coveralls issue, so that we don't forget about it.

File: .github/workflows/gating.yaml [hunk 10]
```diff
@@ -39,7 +39,8 @@ jobs:
       run: |
         pip3 install --upgrade pip
         pip3 install --upgrade setuptools
-        pip3 install --upgrade coveralls==3.2.0
+        pip3 install --upgrade coverage[toml]
```
Comment (eskultety) [inline .github/workflows/gating.yaml:42]: **commit message**

> Note: coveralls is an abandonware now, hasn't been touched
for a couple of years

A reference for such a strong statement would be nice: e.g. this discussion: https://github.com/TheKevJames/coveralls-python/issues/393

File: .github/workflows/gating.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (eskultety) [reply .github/workflows/gating.yaml:43]: linking it here just for completeness: https://github.com/containerbuildsystem/cachi2/issues/498

File: cachi2/core/rooted_path.py [hunk 11]
```diff
@@ -70,8 +70,9 @@ def subpath_from_root(self) -> Path:
         return self._path.relative_to(self._root)
 
     def __eq__(self, other: object) -> bool:
-        # Note: __eq__ is implemented mainly for unit tests that do assert_called_with() or similar
         if not isinstance(other, RootedPath):
+            # NotImplemented is a special value which should be returned by the binary special methods
```
Comment (eskultety) [inline cachi2/core/rooted_path.py:74]: You omitted the important part on which binary methods - **rich comparison methods**.

While at that, you could also tweak the commit subject ever so slightly to say that we're adding ignore patterns rather than increasing the overall coverage in any way as the way it's written now can lead to misunderstandings.

**Review by eskultety — COMMENTED** [review]
@slimreaper35 one more tiny nitpick and we're good to go provided the openshift cluster will let us :(

File: cachi2/core/rooted_path.py (hunk 11 above)
Comment (slimreaper35) [reply cachi2/core/rooted_path.py:74]: Fixed; I copied one more line from the docs for further explanation.

**Review by eskultety — APPROVED** [review]
